Feat #22 [EntityControlPanel] 3D 모델 생성 및 제스쳐 1차 구현

yunsly opened the pull request:
## 📕 Issue Number
#22 
#23
#26 

1. 아직 UI단과 연결하지 않음.
2. 실기기 테스트 필요

의 이유들로 추후 Close 하겠습니다.

## 📙 작업 내역

> 구현 내용 및 작업 했던 내역

**1. ARSessionController**
    - ARSession run / stop
    - 정적 인스턴스로 구현하여 앱 전역에서 접근 가능하나 생성은 private (Singleton)
    - 0.1초마다 deviceTransform 을 업데이트
    
**2. HeadPose** 
    - 마찬가지로 정적 인스턴스로 구현하여 앱 전역에서 접근 가능하나 생성은 private (Singleton)
    - ARSessionController 의 deviceTransform 을 계산한 transform 만 가지고 있음. 
    
**3. EntityPlacementService**
    - AnchorServicing 프로토콜 구현
    - placeAnchorInFront 함수로 바라보는 방향의 AnchorEntity 반환
    - attach 함수에서 EntityID 로 Entity 로드, 제스쳐 붙인 후 Anchor 에 부착
    - detach 함수에서 EntityID 로 removeFromParent
    

**4. RealityEntityLocator 파일 삭제**

## 📘 작업 유형

- [x] 신규 기능 추가
- [ ] 버그 수정
- [ ] 리펙토링
- [ ] 문서 업데이트


## 📝 PR 특이 사항

> PR을 볼 때 주의깊게 봐야하거나 말하고 싶은 점

- ImmersiveSceneRuntime 에서 ImmersiceSpace 가 open/dismiss 될 때마다 ARSession 도 함께 run/stop 되도록 구현하였는데, 차후 entity 를 생성해야할 때만 run 하도록 이동하면 리소스를 줄일 수 있지 않을까 싶습니다.
- 구조를 이렇게 나누는 것이 맞을지 모르겠어서 일단은 최소한으로 나누어 PR 올립니다. 피드백 부탁드려요 빠르게 반영하겠습니다!
- 커밋에 이슈넘버 잘못 적었어요 (27이 아닌 #22 임) 죄송합니다 ㅠ.ㅠ
- 현재 데이터와 연결되지 않아서 차후 UI단이 완성되면 연결 후 Close 하도록 하겠습니다!

<br><br>

yunsly committed:
feat: #27 3D 모델 추가를 위한 ARSessionController 구현

- ARSessionController : ARSession 열고 주기적으로 deviceTransform 업데이트, 정적 인스턴스 구현
- HeadPose : deviceTransform 받아 방향 계산, 싱글톤 구현
- ImmersiveSceneRuntime 에 등록 (추후 수정 필요)

yunsly committed:
feat: #27 3D 모델 추가를 위한 EntityPlacementService 등록

- attach / detach 함수 구현
- entityID 로 entity 로드 -> gesture 부착 후 anchor에 붙임

yunsly committed:
chore: RealityEntityLocator 파일 삭제

yunsly committed:
feat: #26 3D 모델 삭제 함수 구현

- 기존 코드 로직 개선

yunsly committed:
feat: #26 선택한 3D 모델 삭제

- 조작하기 위한 Entity Runtime 프로퍼티로 등록
- detach 함수 수정

hgkim215 commented on Hippo/Features/Vision/Runtime/ARSessionController.swift:
```diff
@@ -0,0 +1,65 @@
+        }
+        
+        // 디바이스 위치 0.1초 간격으로 업데이트
+        timer = Timer.scheduledTimer(withTimeInterval: 0.1, repeats: true) { _ in
```
0.1초 마다 함수가 불러와지는 구조 같은데 좀 더 리소스를 적게 먹게 하는 방향으로 개선할 수 있는 방법이 있나요?

yunsly replied:
요게 특이사항에도 적어둔건데, 나중에 panel 연결 된 이후 entity를 생성해야할 때에만 ARSession 열고 하면 그 때에만 백그라운드로 돌아가게 최적화시킬 수 있을 것 같아요!

hgkim215 commented on Hippo/Features/Vision/Runtime/ARSessionController.swift:
```diff
@@ -0,0 +1,65 @@
+        timer = Timer.scheduledTimer(withTimeInterval: 0.1, repeats: true) { _ in
+            Task.detached {
+                let now = CACurrentMediaTime()
+                if let dev = await self.worldTracking!.queryDeviceAnchor(atTimestamp: now) {
```
저도 !(Force Unwrapping)은 최대한 지향하고 차라리 nil 일때의 기본값을 설정해서 처리해주는게 더 좋다고 배웠습니다!

yunsly replied:
오케 수정해두겠습니다!

hgkim215 commented on Hippo/Features/Vision/Runtime/ARSessionController.swift:
```diff
@@ -0,0 +1,65 @@
+@MainActor
+@Observable
+final class ARSessionController {
+    static let shared = ARSessionController()
```
Singleton 패턴을 사용해야만 하는 이유가 있을까요?

yunsly replied:
앱 전역에서 하나의 인스턴스를 접근해야 세션 관리가 편할 것 같아서 싱글톤을 사용하였습니다..!

hgkim215 commented on Hippo/Features/Vision/Runtime/HeadPose.swift:
```diff
@@ -0,0 +1,30 @@
+class HeadPose {
+    
+    // Singleton
+    static let instance = HeadPose()
```
Singleton 패턴과 Observable 매크로를 함께 사용해야 하는 이유가 궁금합니다!

yunsly replied:
headpose 인스턴스가 여러개일 필요가 없고, 인스턴스 하나를 갖고 있으면서 외부에서는 내부 업데이트되는 transform 만 접근하면 되는 구조여서 싱글톤으로 구현했고, 지금 생각해보니 딱히 Observable 일 필요가 없네요 프로퍼티로만 접근해서리... 굿굿 수정하겠습니다!

yunsly commented on Hippo/Features/Vision/Runtime/ARSessionController.swift:
```diff
@@ -0,0 +1,65 @@
+        timer?.invalidate()
+        timer = nil
+        Task { [session] in
+            session!.stop()
```
if let 구문으로 수정

hgkim215 reviewed (approved):
일단 머지하고 나중에 코드 전체적으로 들어갔을때 한번 리팩토링해보시죠? 👍 
고생하셨습니다 😃

yunsly committed:
refactor: PR 피드백 수정

- 필요없는 do 구문 삭제
- 강제 언래핑 코드 수정
- 이벤트 구독 변수 추가하여 리소스 관리, Immersive 환경 나갈 시 구독 함께 정리
